datafusion/physical-expr/src/expressions/case/literal_lookup_table/mod.rs

```diff
@@ -0,0 +1,327 @@
+            // All WHEN literals were nulls, so cannot use optimization
+            //
+            // instead, another optimization would be to go straight to the ELSE clause
+            return None;
+        }
+
+        // Keep only the first occurrence of each when literal (as the first match is used)
+        // and remove nulls (as they cannot be matched - case NULL WHEN NULL THEN ... ELSE ... END always goes to ELSE)
+        let (when, then): (Vec<ScalarValue>, Vec<ScalarValue>) = {
+            let mut map = IndexMap::with_capacity(body.when_then_expr.len());
+
+            for (when, then) in when_then_exprs_scalars.into_iter() {
+                // Don't overwrite existing entries as we want to keep the first occurrence
+                if !map.contains_key(&when) {
+                    map.insert(when, then);
+                }
+            }
+
+            map.into_iter().unzip()
+        };
+
+        let else_value: ScalarValue = if let Some(else_expr) = &body.else_expr {
+            let literal = else_expr.as_any().downcast_ref::<Literal>()?;
+
+            literal.value().clone()
+        } else {
+            let Ok(null_scalar) = ScalarValue::try_new_null(&then[0].data_type()) else {
+                return None;
+            };
+
+            null_scalar
+        };
+
+        {
+            let when_data_type = when[0].data_type();
+
+            // If not all the WHEN literals are the same data type we cannot use this optimization
+            if when.iter().any(|l| l.data_type() != when_data_type) {
+                return None;
+            }
+        }
+
+        {
+            let data_type = then[0].data_type();
+
+            // If not all the then and the else literals are the same data type we cannot use this optimization
+            if then.iter().any(|l| l.data_type() != data_type) {
+                return None;
+            }
+
+            if else_value.data_type() != data_type {
+                return None;
+            }
+        }
+
+        let then_and_else_values = ScalarValue::iter_to_array(
+            then.iter()
+                // The else is in the end
+                .chain(std::iter::once(&else_value))
+                .cloned(),
+        )
+        .ok()?;
+        // The else expression is in the end
+        let else_index = then_and_else_values.len() as u32 - 1;
+
+        let lookup = try_creating_lookup_table(when).ok()?;
+
+        Some(Self {
+            lookup,
+            then_and_else_values,
+            else_index,
+        })
+    }
+
+    pub(in super::super) fn map_keys_to_values(
+        &self,
+        keys_array: &ArrayRef,
+    ) -> datafusion_common::Result<ArrayRef> {
+        let take_indices = self
+            .lookup
+            .map_to_when_indices(keys_array, self.else_index)?;
+
+        // Zero-copy conversion
+        let take_indices = UInt32Array::from(take_indices);
+
+        // An optimize version would depend on the type of the values_to_take_from
```

Comment:
I believe the `take` kernel already implements the optimization this comment is referring to (this is one of the benefits of re-using the arrow kernels). Thus I think this comment no longer applies